toolkit/tools/internal/pkggraph/graphprinter.go

```diff
@@ -0,0 +1,142 @@
+}
+
+type graphPrinterConfigModifier func(*graphPrinterConfig)
+
+// loggerOutputWrapper is a wrapper around logrus.Logger to implement the io.StringWriter interface.
+type loggerOutputWrapper struct {
+	logLevel logrus.Level
+}
+
+// Write implements the io.StringWriter interface.
+func (d *loggerOutputWrapper) WriteString(s string) (int, error) {
+	logger.Log.Log(d.logLevel, s)
+	return len(s), nil
+}
+
+// NewGraphPrinter creates a new GraphPrinter.
+// It accepts a variadic number of 'GraphPrinter*' modifiers to customize the printer's behavior.
+// The default settings are:
+// - Indent string: "  " (2 spaces)
+// - Output: logrus logger on debug level
+func NewGraphPrinter(configModifiers ...graphPrinterConfigModifier) *GraphPrinter {
+	config := &graphPrinterConfig{
+		indentString: "  ",
+		output: &loggerOutputWrapper{
+			logLevel: logrus.DebugLevel,
+		},
+	}
+
+	for _, modifier := range configModifiers {
+		modifier(config)
+	}
+
+	return &GraphPrinter{
+		graphPrinterConfig: *config,
+	}
+}
+
+// GraphPrinterIndentString is a config modifier passed to the graph printer's constructor
+// to define the string used for indentation in the graph printer.
+func GraphPrinterIndentString(indentString string) graphPrinterConfigModifier {
+	return func(c *graphPrinterConfig) {
+		c.indentString = indentString
+	}
+}
+
+// GraphPrinterOutput is a config modifier passed to the graph printer's constructor
+// to define the output writer for the graph printer.
+func GraphPrinterOutput(output io.StringWriter) graphPrinterConfigModifier {
+	return func(c *graphPrinterConfig) {
+		c.output = output
+	}
+}
+
+// GraphPrinterLogOutput is a config modifier passed to the graph printer's constructor
+// making the printer's output be logged at the specified log level.
+func GraphPrinterLogOutput(logLevel logrus.Level) graphPrinterConfigModifier {
+	return func(c *graphPrinterConfig) {
+		c.output = &loggerOutputWrapper{
+			logLevel: logLevel,
+		}
+	}
+}
+
+// Print prints the graph in the following manner:
+// - Each line represents a node in the graph.
+// - Each node is represented by its friendly name.
+// - Children of each node have an indentation level based on their depth in the graph.
+func (g GraphPrinter) Print(graph *PkgGraph, rootNode *PkgNode) error {
+	var dfsPrint func(*PkgNode) error
+
+	if graph == nil {
+		return fmt.Errorf("graph is nil")
+	}
+
+	if rootNode == nil {
+		return fmt.Errorf("root node is nil")
+	}
+
+	if !graph.HasNode(rootNode) {
+		return fmt.Errorf("root node '%s' not found in the graph", rootNode.FriendlyName())
+	}
+
+	level := 0
+	// Use a set to keep track of seen nodes to avoid infinite loops.
+	seenNodes := make(map[*PkgNode]bool)
+
+	// Walking the graph manually to be able to track the depth level.
+	dfsPrint = func(node *PkgNode) error {
+		if node == nil || seenNodes[node] {
+			return nil
+		}
+
+		line := fmt.Sprintf("%s%s\n", strings.Repeat(string(g.indentString), level), node.FriendlyName())
```

Comment:
Highly recommend not taking too much inspiration from this code, but ... https://github.com/microsoft/azurelinux/blob/7e483a893e4413d0002bfc3080f7175e4b371289/toolkit/tools/pkg/depsearch/depsearch.go#L283

tldr, I was doing the same and settled on something that looks like
```
pkg
|--> child
|       |--> child2
|
|--> child 3
|
```

Another consideration is duplicates (ie a child can have multiple parrents). I added a flag to omit duplicates which cut down the graph size significantly.

Reply:
That's a good point about limiting the size of the output. In interest of limiting the size of the change, I'm going to skip adding that option now. But I hope that it shouldn't be too troublesome to extend the printer in the future, if needed.

Reply:
I've modified the way the graph is printer now, though, by using the ["gtree" module](https://github.com/ddddddO/gtree). I hope it is more readable now.